File: src/UpToU.Core/Entities/Story.cs
namespace UpToU.Core.Entities;

public class Story
{
    public int Id { get; set; }
    public string Title { get; set; } = string.Empty;
    public string? Slug { get; set; }
    public string? Description { get; set; }
    public string? Excerpt { get; set; }
    public string? CoverImageUrl { get; set; }
    public string? AuthorName { get; set; }
    public bool IsFeatured { get; set; } = false;
    public DateTime? PublishDate { get; set; }
    public bool IsPublish { get; set; } = false;
    public bool IsDeleted { get; set; } = false;
    public DateTime CreatedOn { get; set; } = DateTime.UtcNow;
    public DateTime? ModifiedOn { get; set; }
    public string? CreatedBy { get; set; }
    public string? ModifiedBy { get; set; }

    public int CategoryId { get; set; }
    public Category Category { get; set; } = null!;

    public int ViewCount { get; set; } = 0;

    /// <summary>"Article" | "Interactive"</summary>
    public string StoryType { get; set; } = "Article";

    // ── Workflow ─────────────────────────────────────────────────────────────
    /// <summary>Id of the user (Contributor/Admin) who created this story.</summary>
    public string? AuthorId { get; set; }

    /// <summary>Workflow status: Draft | Submitted | Approved | Published | Rejected.</summary>
    public string Status { get; set; } = StoryStatus.Draft;

    public DateTime? SubmittedAt  { get; set; }
    public string?   ReviewedBy   { get; set; }
    public DateTime? ReviewedAt   { get; set; }
    public string?   RejectionReason { get; set; }

    /// <summary>Optional supervisor specifically assigned by the contributor to review this story.</summary>
    public string? AssignedSupervisorId { get; set; }

    // ── Score completion ──────────────────────────────────────────────────────
    /// <summary>The score type that gates story completion. When a user's score for this type reaches <see cref="MaxScoreValue"/>, the story ends and the user is marked as having finished it.</summary>
    public int? MaxScoreTypeId { get; set; }
    public CategoryScoreType? MaxScoreType { get; set; }

    /// <summary>Maximum score value for <see cref="MaxScoreTypeId"/>. Null means no score-based completion gate.</summary>
    public int? MaxScoreValue { get; set; }

    public ICollection<Tag> Tags { get; set; } = new List<Tag>();
    public ICollection<StoryDetail> StoryDetails { get; set; } = new List<StoryDetail>();
    public ICollection<Comment> Comments { get; set; } = new List<Comment>();
    public ICollection<Reaction> Reactions { get; set; } = new List<Reaction>();
    public ICollection<StoryVote> StoryVotes { get; set; } = new List<StoryVote>();
    public ICollection<StoryTranslation> Translations { get; set; } = new List<StoryTranslation>();
}

File: client/src/types/story.ts
import type { VoteType } from './vote';

export interface Tag {
  id: number;
  name: string;
}

export interface StoryDetail {
  id: number;
  storyId: number;
  revision: number;
  isPublish: boolean;
  content: string | null;
  wordCount: number;
  changeNotes: string | null;
  scoreWeight: number;
  scoreWeightHistory: number[];
  savePath: string;
  createdOn: string;
  createdBy: string | null;
}

export interface Story {
  id: number;
  title: string;
  slug: string | null;
  description: string | null;
  excerpt: string | null;
  coverImageUrl: string | null;
  authorName: string | null;
  isFeatured: boolean;
  publishDate: string | null;
  isPublish: boolean;
  isDeleted: boolean;
  categoryId: number;
  categoryTitle: string;
  createdOn: string;
  modifiedOn: string | null;
  createdBy: string | null;
  modifiedBy: string | null;
  tags: Tag[];
  latestDetail: StoryDetail | null;
  viewCount: number;
  upvoteCount: number;
  downvoteCount: number;
  currentUserVote: VoteType | null;
  isBookmarked: boolean;
  storyType: string;
}
